### PR TITLE
[WIP] Scheduler

### PR DESCRIPTION
**Pull request by mandelsoft**
**What this PR does / why we need it**:

Reopen gardener#1567

The new scheduling method Selector allows to make use of the new label selector field
providers to select seeds running on different providers depending of the provider list list and matching labels. It prefers matching seeds if the load is the same.

**Special notes for your reviewer**:

added unit tests and rebased old pull request.
Default strategy has been reset to the original one

**Release note**:

```improvement operator
If automatic cross provider scheduling is desired, it  is possible know to specify a new seed selector field `providers` for a cloud profile to enable scheduling on seeds running on different providers.
It is observed if the scheduling method `Selector` has been chosen, which is the new default now.
```

### CODE REVIEW

File: pkg/scheduler/controller/shoot/scheduler_control.go [hunk 1]
```diff
@@ -175,15 +175,37 @@ func determineSeed(shoot *gardencorev1beta1.Shoot, seedLister gardencorelisters.
 		return nil, err
 	}
 
+	logger.Logger.Debugf("strategy: %s\n", strategy)
+	for _, s := range seedList {
```
Comment (danielfoehrKn) [inline pkg/scheduler/controller/shoot/scheduler_control.go:179]: Should that be in the productive code?

File: pkg/apis/core/v1alpha1/types_cloudprofile.go [hunk 2]
```diff
@@ -83,6 +83,19 @@ type CloudProfileSpec struct {
...
+type SeedSelector struct {
+	// LabelSelector is optional and can be used to select seeds by their label settings
+	// +optional
+	*metav1.LabelSelector `json:",inline,omitempty"`
+	// Providers is optional and can be used by restricting seeds by their provider type. '*' can be used to enable seeds regardless of their provider type.
+	// +optional
+	Providers []string `json:"providers,omitempty"`
+	// Seeds is optional and can be used by restricting seeds by their name. '*' can be used to enable seeds regardless of their name.
+	// +optional
+	Seeds []string `json:"seeds,omitempty"`
```
Comment (timebertt) [inline pkg/apis/core/v1alpha1/types_cloudprofile.go:96]: What would be the usecase of selecting Seeds by their name?
This seems cumbersome, I cannot think about a place in kubernetes where you select something by its name.

**Comment by mandelsoft** [comment]
We decided to redesign the complete scheduling mechanism to correctly handle field selectors and label selectors in combination with scheduling policies assigned to cloud profiles.
The policies should be able to handle cross platform assignments correctly, especially we need a metric to handle the nearest seed feature in all scenarios, including complete on-prem scenarios based on openstack and vmware using completely arbitrary region names.
So the actual proposal is obsolete.

**Comment by rfranzke** [comment]
In another round of discussions we have talked about additional scheduling requirements like EU access-only, and similar. We talked about potential reuse of the taints/tolerations concept for nodes/pods in k/k for our seeds/shoots. Anyways, we agreed on spending some time to rethink the whole design to support these various scenarios with a flexible and expressive API. @mandelsoft wants to file a GEP with his initial ideas once ready. Hence, we'll close this PR for now as we won't finish/merge it.

**Comment by vlerenc** [comment]
OK, but what's that? Can't we first have a discussion around sample cloud profiles and how we envision the future configuration. That here is too much for me. If we want to discuss our future specs I am in, but if everything is already in code, it feels like it's already too late.

**Pull request closed without merging**